executor/new_executor.go

```diff
@@ -149,12 +217,86 @@ func (e *HashJoinExec) prepare() error {
 		}
 	}
 
+	e.resultRows = make(chan *Row, e.concurrency*1000)
+	e.resultErr = make(chan error, 1)
+
+	e.wg = sync.WaitGroup{}
+	for i := 0; i < e.concurrency; i++ {
+		e.wg.Add(1)
+		go e.doJoin(i)
+	}
+	go e.closeChanWorker()
+
 	e.prepared = true
 	return nil
 }
 
-func (e *HashJoinExec) constructMatchedRows(bigRow *Row) (matchedRows []*Row, err error) {
-	hasNull, hashcode, err := getHashKey(e.bigHashKey, bigRow, e.targetTypes)
+func (e *HashJoinExec) closeChanWorker() {
+	e.wg.Wait()
+	close(e.resultRows)
+}
+
+// doJoin does join job in one goroutine.
+func (e *HashJoinExec) doJoin(idx int) {
+	for {
+		var (
+			bigRows []*Row
+			ok      bool
+			err     error
+		)
+		select {
+		case bigRows, ok = <-e.bigTableRows[idx]:
+		case err = <-e.bigTableErr:
```

Comment:
See line 365.
If any error found in  resultErr, e.finished will be set to true.